code/modules/reagents/Chemistry-Holder.dm

```diff
@@ -599,18 +599,49 @@ trans_to_atmos(var/datum/gas_mixture/target, var/amount=1, var/multiplier=1, var
 
 	for (var/datum/reagent/R in reagent_list)
 		if (R.id == reagent)
-			if(!R.on_removal(amount))
-				return 0 //handled and reagent says fuck no
-			R.volume -= amount
-			update_total()
-			if(!safety)//So it does not handle reactions when it need not to
-				handle_reactions()
-			if(my_atom)
-				my_atom.on_reagent_change()
-			return 0
+			return remove_that_reagent(R, amount, safety)
+	return 1
+
+/datum/reagents/proc/remove_reagents(var/list/reagent_list, var/amount, var/safety)
+	if(!isnum(amount))
+		return 1
 
+	for(var/id in reagent_list)
+		if(has_reagent(id, amount))
+			remove_reagent(id, amount, safety)
 	return 1
 
+/datum/reagents/proc/remove_reagent_by_type(var/reagent_type, var/amount, var/safety)
+	if(!isnum(amount))
+		return 1
+
+	var/datum/reagent/R = get_reagent_by_type(type, amount)
+	if(R)
+		remove_that_reagent(R, amount, safety)
+
+/datum/reagents/proc/remove_reagents_by_type(var/list/reagent_types, var/amount, var/safety)
+	if(!isnum(amount))
+		return 1
+
+	for(var/datum/reagent/R in reagent_list)
+		if(is_type_in_list(R, reagent_types))
```

Comment:
I might be saying something dumb, but what about `R.type in reagent_types` - would it be faster or no

Reply:
It would be faster, but not suitable for this proc.

R.type in reagent_types would check strictly ==, whereas is_type_in_list would be doing istype(R, reagent_types[i]) until success